internal/mode/static/nginx/modules/src/httpmatches.js

```diff
@@ -49,32 +48,37 @@ function redirect(r) {
   r.internalRedirect(match.redirectPath);
 }
 
-function extractMatchesFromRequest(r) {
-  if (!r.variables[MATCHES_VARIABLE]) {
-    throw Error(
-      `cannot redirect the request; the variable ${MATCHES_VARIABLE} is not defined on the request object`,
-    );
-  }
-
-  let matches;
-
+function internalRedirect(r, matchList) {
+  // matchList is a list of http matches in order of precedence.
+  // We will accept the first match that the request satisfies.
+  // If there's a match, redirect request to internal location block.
+  // If an exception occurs, return 500.
+  // If no matches are found, return 404.
+  let match;
   try {
-    matches = JSON.parse(r.variables[MATCHES_VARIABLE]);
+    match = findWinningMatch(r, matchList);
   } catch (e) {
-    throw Error(
-      `cannot redirect the request; error parsing ${r.variables[MATCHES_VARIABLE]} into a JSON object: ${e}`,
-    );
+    r.error(e.message);
+    r.return(HTTP_CODES.internalServerError);
+    return;
   }
 
-  if (!Array.isArray(matches)) {
-    throw Error(`cannot redirect the request; expected a list of matches, got ${matches}`);
+  if (!match) {
+    r.return(HTTP_CODES.notFound);
+    return;
   }
 
-  if (matches.length === 0) {
```

Comment:
Feels like we should still be doing similar validation as before.

Reply:
will add the same validation to the match list